source/docs/software/wpilib-tools/shuffleboard/custom-widgets/creating-plugins.rst

```diff
@@ -0,0 +1,101 @@
+Creating a Plugin
+=================
+
+Overview
+--------
+Plugins provide the ability to create custom widgets, layouts, data sources/types, and custom themes. Shuffleboard provides the following plugins:
+
+- NetworkTables Plugin: To connect to data published over NetworkTables
+- Base Plugin: To display custom FRC data types in custom widgets
+- CameraServer Plugin: To view streams from the CameraServer
+
+Create a Custom Plugin
+----------------------
+In order to define a plugin, the plugin class must be a subclass of `edu.wpi.first.shuffleboard.api.Plugin <https://github.com/wpilibsuite/shuffleboard/blob/master/api/src/main/java/edu/wpi/first/shuffleboard/api/plugin/Plugin.java>`_ or one of its subclasses. An example of a plugin class would be as following.
+
+.. tabs::
+
+   .. code-tab:: java
+
+      import edu.wpi.first.shuffleboard.api.plugin.Description;
+      import edu.wpi.first.shuffleboard.api.plugin.Plugin;
+
+      @Description(group = "com.example", name = "MyPlugin", version = "1.2.3", summary = "An example plugin")
+      public class MyPlugin extends Plugin {
+
+      }
+
+More resources on how these attributes are used, including version numbers Shuffleboard's plugin dependencies can be found `here <https://semver.org/>`_.
+
+Note the ``@Description`` annotation which is needed to tell the plugin loader the properties of the custom plugin class.
+Plugin classes are permitted to have a defualt constructor but it cannot take any arguments.
+
+Building plugin
+---------------
+Plugins require the usuage of the `Shuffleboard API Library <https://frcmaven.wpi.edu/artifactory/release/edu/wpi/first/shuffleboard/api/>`_. These dependencies can be resolved in the
+``build.gradle`` file or using maven. The dependencies would be as follows:
+
+For Gradle:
+
+.. code-block:: groovy
+
+   repositories {
+      mavenCentral()
+      maven{ url "https://frcmaven.wpi.edu:443/artifactory/release" }
+   }
+
+   dependencies {
+      compileOnly("edu.wpi.first.shuffleboard", "api", "2020.+")
+      compileOnly("edu.wpi.first.shuffleboard.plugin", "networktables", "2020.+")
+   }
+
+For Maven:
+
+ .. code-blocl:: groovy
+
+Plugins are allowed to have dependencies on other plugins and libraries, however, they must be included correctly in the maven or gradle build file.
+When having Plugin dependencies to other plugins, it is good practise to define those dependencies so the plugin does not load when the dependencies do not load as well.
+This can be done using the ``@Requires`` annotation as shown below:
```

Comment:
```suggestion
Plugins are allowed to have dependencies on other plugins and libraries, however, they must be included correctly in the maven or gradle build file. When having Plugin dependencies to other plugins, it is good practice to define those dependencies so the plugin does not load when the dependencies do not load as well. This can be done using the ``@Requires`` annotation as shown below:
```